oauth_tools/external_idp.py

```diff
@@ -0,0 +1,206 @@
+#!/usr/bin/env python3
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+import abc
+import logging
+from os.path import join
+from time import sleep
+from typing import List, Optional
+
+import requests
+from lightkube import Client, KubeConfig, codecs
+from lightkube.core.exceptions import ApiError, ObjectDeleted
+from lightkube.resources.apps_v1 import Deployment
+from lightkube.resources.core_v1 import Namespace, Pod, Service
+from pytest_operator.plugin import OpsTest
+from requests.exceptions import RequestException
+
+from oauth_tools.constants import (
+    DEX_CLIENT_ID,
+    DEX_CLIENT_SECRET,
+    DEX_MANIFESTS,
+    EXTERNAL_USER_EMAIL,
+    EXTERNAL_USER_PASSWORD,
+    KUBECONFIG,
+)
+
+logger = logging.getLogger(__name__)
+
+
+class ExternalIdpManager(abc.ABC):
+    """Abstract class for managing lifecycle for an external IdP."""
+
+    @property
+    def client_id(self) -> str:
+        """The client_id of a registered client."""
+        ...
+
+    @property
+    def client_secret(self) -> str:
+        """The client_sercet of a registered client."""
+        ...
+
+    @property
+    def user_email(self) -> str:
+        """The test user's email."""
+        ...
+
+    @property
+    def user_password(self) -> str:
+        """The test user's password."""
+        ...
+
+    @property
+    def issuer_url(self) -> str:
+        """The provider's issuer URL."""
+        ...
+
+    @abc.abstractmethod
+    def create_idp_service(self) -> None:
+        """Deploy and configure the idp service."""
+        ...
+
+    @abc.abstractmethod
+    def wait_until_is_ready(self) -> None:
```

Comment:
Just a **NONBLOCKING** question: does this method need to be exposed? from the only subclass, it seems that no client cares whether should wait or not.

Reply:
You are right, the rest of the methods that are exposed all wait for their effects to take place. I thought that it may be useful in some cases, but I will make it private for now